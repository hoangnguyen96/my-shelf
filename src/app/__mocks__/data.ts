export const DATA_BOOKS = [
  {
    title: "Don't Make Me Think",
    author: "Steve Krug",
    category: "SelfHelpBook",
    imageUrl: "https://i.ibb.co/QbQgVtG/book1.png",
    description: "description will update later",
    status: false,
    publicationYear: 2000,
    rating: 4.5,
    createdDate: "3 Sep 2024 9:07 PM",
    edition: "first",
    id: "1",
  },
  {
    title: "Rich Dad Poor Dad",
    author: "Robert T. Kiyosaki",
    category: "SelfHelpBook",
    imageUrl: "https://i.ibb.co/9qHTtQF/book2.png",
    description:
      "Steve Krug is a usability consultant who has more than 30 years of experience as a user advocate for companies like Apple, Netscape, AOL, Lexus, and others. Based in part on the success of his first book, Don't Make Me Think, he has become a highly sought-after speaker on usability design.",
    status: false,
    publicationYear: 1997,
    rating: 4.5,
    createdDate: "3 Sep 2024 8:47 PM",
    edition: "second",
    id: "2",
  },
  {
    title: "The Design of Everyday Things",
    author: "Don Norman",
    category: "TextBook",
    imageUrl: "https://i.ibb.co/LnSBcCM/book5.png",
    description:
      "Steve Krug is a usability consultant who has more than 30 years of experience as a user advocate for companies like Apple, Netscape, AOL, Lexus, and others. Based in part on the success of his first book, Don't Make Me Think, he has become a highly sought-after speaker on usability design.",
    status: false,
    publicationYear: 1988,
    rating: 4.5,
    createdDate: "3 Sep 2024 8:51 PM",
    edition: "fourth",
    id: "3",
  },
];

export const DATA_USER = [
  {
    username: "admin",
    email: "admin@gmail.com",
    password: "$2a$10$MwSZ9vWSX7DVJwJ4SF5Mzuj2zMFjY52XM9L11gGG0j79gybowaLfq",
    isAdmin: true,
    phone: "",
    bio: "",
    avatar: "https://i.ibb.co/RHMqQGr/man-1.png",
    favorites: ["3", "2", "1", "13", "5"],
    shelfBooks: ["1", "6"],
    userId: "3733403",
    id: "1",
  },
  {
    username: "hung",
    email: "hung1@gmail.com",
    password: "$2a$10$Mt6ruwOzt3nzX1.HpSt.OOzut.teXmLE74WclmkP.Nn.gzVQrlqri",
    isAdmin: false,
    phone: "",
    bio: "",
    avatar: "https://i.ibb.co/88X1WfZ/avatar-default.png",
    favorites: ["1"],
    shelfBooks: ["2", "3", "1"],
    userId: "8857134",
    id: "2",
  },
];
