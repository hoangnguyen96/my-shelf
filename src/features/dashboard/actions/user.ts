import { API_ROUTES, MESSAGES } from "@app/constants";
import { User } from "@app/models";
import { api } from "@app/services";

export const getAllUser = async () => {
  try {
    const data = await api.get<User[]>(API_ROUTES.USER);

    return data || [];
  } catch (error) {
    if (error instanceof Error) {
      return error.message;
    }
    return MESSAGES.RESPONSE_ERROR;
  }
};

export const getUserById = async (id: string) => {
  try {
    const data = await api.get<User[]>(`${API_ROUTES.USER}?userId=${id}`);

    return data[0];
  } catch (error) {
    if (error instanceof Error) {
      return error.message;
    }
    return MESSAGES.RESPONSE_ERROR;
  }
};

export const getUserByEmail = async (email: string) => {
  try {
    const user = (await api.get(`${API_ROUTES.USER}?email=${email}`)) as User[];

    return user;
  } catch (error) {
    if (error instanceof Error) {
      return error.message;
    }
    return MESSAGES.RESPONSE_ERROR;
  }
};

export const updateUserById = async (id: string, payload: Partial<User>) => {
  return await api.put(`${API_ROUTES.USER}/${id}`, payload);
};

export const addUser = async (payload: Partial<User>) => {
  return await api.post(API_ROUTES.USER, payload);
};
