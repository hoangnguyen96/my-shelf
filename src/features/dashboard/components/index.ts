export * from "./home-list";
export * from "./search-list";
export * from "./my-book-shelf";
export * from "./my-book-shelf-favorites";
export * from "./preview-book-details";
export * from "./profile-update";
export * from "./contribute";
export * from "./contribute-update";
export * from "./contribute-complete";
export * from "./contribute-list";
