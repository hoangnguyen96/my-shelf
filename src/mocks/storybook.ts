export const mockRouter = {
  pathname: "/",
  route: "/",
  query: {},
  asPath: "/",
  basePath: "",
  push: () => Promise.resolve(true),
  replace: () => Promise.resolve(true),
  reload: () => {},
  back: () => {},
  prefetch: () => Promise.resolve(),
  beforePopState: () => {},
  isFallback: false,
  isLocaleDomain: false,
  isReady: true,
  isPreview: false,
  forward: () => {},
  events: {
    on: () => {},
    off: () => {},
    emit: () => {},
  },
};

export const mockSession = {
  user: {
    isAdmin: true,
    email: "admin@gmail.com",
    id: "3733403",
    name: "admin",
    image: "https://i.ibb.co/RHMqQGr/man-1.png",
  },
  expires: "2024-12-31T23:59:59.999Z",
};
