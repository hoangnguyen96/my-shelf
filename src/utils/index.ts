export * from "./book";
export * from "./date";
export * from "./error";
export * from "./search";
export * from "./user";
export * from "./validators";
