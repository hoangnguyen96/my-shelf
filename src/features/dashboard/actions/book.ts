import { BookType } from "@app/models";
import { API_ROUTES, MESSAGES } from "@app/constants";
import { api } from "@app/services";
import { dividePaginationBooks, getThreeTopBook } from "@app/utils";

export const getAllBook = async () => {
  try {
    const data = await api.get<BookType[]>(API_ROUTES.BOOKS);

    return data;
  } catch (error) {
    if (error instanceof Error) {
      return error.message;
    }
    return MESSAGES.RESPONSE_ERROR;
  }
};

export const getTopThreeBook = async () => {
  try {
    const data = await api.get<BookType[]>(API_ROUTES.BOOKS, {
      next: { revalidate: 60 },
    });
    const result = getThreeTopBook(data);

    return result;
  } catch (error) {
    if (error instanceof Error) {
      return error.message;
    }
    return MESSAGES.RESPONSE_ERROR;
  }
};

export const getPaginatedBook = async (params?: string) => {
  try {
    const data = await api.get<BookType[]>(
      `${API_ROUTES.BOOKS}?${params || ""}`,
      {
        cache: "no-store",
      }
    );
    const result = dividePaginationBooks(data);

    return result;
  } catch (error) {
    if (error instanceof Error) {
      return error.message;
    }
    return MESSAGES.RESPONSE_ERROR;
  }
};

export const getTwelveItemBook = async (params?: string) => {
  try {
    const data = await api.get<BookType[]>(
      `${API_ROUTES.BOOKS}?${params || ""}page=1&limit=12`,
      { next: { revalidate: 3600 } }
    );

    return data;
  } catch (error) {
    if (error instanceof Error) {
      return error.message;
    }
    return MESSAGES.RESPONSE_ERROR;
  }
};

export const getBookByParams = async (params: string) => {
  try {
    const data = await api.get<BookType[]>(`${API_ROUTES.BOOKS}?${params}`);

    return data;
  } catch (error) {
    if (error instanceof Error) {
      return error.message;
    }
    return MESSAGES.RESPONSE_ERROR;
  }
};

export const getBookById = async (id: string) => {
  try {
    const data = await api.get<BookType>(`${API_ROUTES.BOOKS}/${parseInt(id)}`);

    return data;
  } catch (error) {
    if (error instanceof Error) {
      return error.message;
    }
    return MESSAGES.RESPONSE_ERROR;
  }
};

export const addBook = async (payload: Partial<BookType>) => {
  return await api.post(API_ROUTES.BOOKS, payload);
};

export const updateBookById = async (
  id: string,
  payload: Partial<BookType>
) => {
  return await api.put(`${API_ROUTES.BOOKS}/${id}`, payload);
};

export const deleteBook = async (id: string) => {
  return await api.delete(`${API_ROUTES.BOOKS}/${id}`);
};
