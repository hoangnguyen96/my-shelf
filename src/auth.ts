import NextAuth from "next-auth";
import { authConfig } from "./auth.config";
import Credentials from "next-auth/providers/credentials";
import { z } from "zod";
import bcrypt from "bcryptjs";
import { MESSAGES } from "./constants";
import { getUserByEmail } from "@app/features/dashboard/actions";
import { User } from "./models";

export const { handlers, auth, signIn, signOut } = NextAuth({
  ...authConfig,
  secret: process.env.AUTH_SECRET,
  providers: [
    Credentials({
      name: "Credentials",
      credentials: {
        email: { label: "Email", type: "text" },
        password: { label: "Password", type: "password" },
      },
      async authorize(credentials) {
        if (!credentials) {
          return null;
        }

        const parsedCredentials = z
          .object({ email: z.string().email(), password: z.string().min(6) })
          .safeParse(credentials);

        if (parsedCredentials.success) {
          const { email, password } = parsedCredentials.data;
          const user = (await getUserByEmail(email)) as User;

          if (!user || !user.password) {
            throw new Error(MESSAGES.LOGIN_NOTFOUND);
          }

          const passwordsMatch = await bcrypt.compare(password, user.password);

          if (passwordsMatch)
            return {
              id: user.userId,
              name: user.username,
              isAdmin: user.isAdmin,
              image: user.avatar,
              email: user.email,
            };
        }

        throw new Error(MESSAGES.INVALID_CREDENTIALS);
      },
    }),
  ],
});
