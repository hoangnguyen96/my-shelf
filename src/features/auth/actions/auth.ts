"use server";

import { AuthError } from "next-auth";
import { MESSAGES } from "@app/constants";
import { User } from "@app/interface";
import { signIn, signOut } from "@app/auth";

export const authenticate = async (formData: Partial<User>) => {
  try {
    await signIn("credentials", formData);
  } catch (error) {
    if (error instanceof AuthError) {
      return MESSAGES.EMAIL_PASSWORD_INVALID;
    }
    throw error;
  }
};

export const logout = async () => {
  await signOut();
};
