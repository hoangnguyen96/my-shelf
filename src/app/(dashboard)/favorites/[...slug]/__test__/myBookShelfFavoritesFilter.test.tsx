import { render, waitFor } from "@testing-library/react";
import { getAllBook, getUserById } from "@app/features/dashboard/actions";
import { DATA_BOOKS, DATA_USER } from "@app/mocks/data";
import * as utils from "@app/utils";
import MyBookShelfFavoritesPage from "../page";
import { auth } from "@app/auth";

jest.mock("next/navigation", () => ({
  useRouter: jest.fn(),
}));

jest.mock("@app/auth", () => ({
  auth: jest.fn(),
}));

jest.mock("@app/features/dashboard/actions", () => ({
  getAllBook: jest.fn(),
  getUserById: jest.fn(),
}));

jest.mock("@app/utils", () => ({
  ...jest.requireActual("@app/utils"),
  filterBooksFavorite: jest.fn(),
  filterBooksFavoriteByParams: jest.fn(),
}));

describe("My Book Shelf Favorites Search", () => {
  const mockBooks = DATA_BOOKS.filter((item) =>
    DATA_USER[0].favorites.includes(item.id)
  );
  const mockBooksByParams = mockBooks.filter((book) =>
    book.title.includes("on")
  );
  const mockSession = {
    user: {
      isAdmin: true,
      email: "admin@gmail.com",
      id: "3733403",
      name: "admin",
      image: "https://i.ibb.co/RHMqQGr/man-1.png",
    },
  };

  beforeEach(() => {
    jest.clearAllMocks();
    (auth as jest.Mock).mockResolvedValue(mockSession);
    (getAllBook as jest.Mock).mockReturnValue({
      data: DATA_BOOKS,
    });
    (getUserById as jest.Mock).mockReturnValue({
      data: DATA_USER[0],
    });
    jest.spyOn(utils, "filterBooksFavorite").mockReturnValue(mockBooks);
    jest
      .spyOn(utils, "filterBooksFavoriteByParams")
      .mockReturnValue(mockBooksByParams);
  });

  it("Should render correctly snapshot", async () => {
    const params = { slug: ["title", "on"] };
    const { container } = render(await MyBookShelfFavoritesPage({ params }));

    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });
});
