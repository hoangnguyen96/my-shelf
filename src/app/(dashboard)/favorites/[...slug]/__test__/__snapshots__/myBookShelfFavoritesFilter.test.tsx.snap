// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`My Book Shelf Favorites Search Should render correctly snapshot 1`] = `
<div>
  <div
    class="css-6q2x0w"
  >
    <div
      class="css-1q7xe83"
    >
      <div
        class="css-dk70ac"
      >
        <img
          alt="Don't Make Me Think"
          data-nimg="1"
          decoding="async"
          fetchpriority="high"
          height="100"
          src="/_next/image?url=https%3A%2F%2Fi.ibb.co%2FQbQgVtG%2Fbook1.png&w=256&q=75"
          srcset="/_next/image?url=https%3A%2F%2Fi.ibb.co%2FQbQgVtG%2Fbook1.png&w=96&q=75 1x, /_next/image?url=https%3A%2F%2Fi.ibb.co%2FQbQgVtG%2Fbook1.png&w=256&q=75 2x"
          style="color: transparent;"
          width="75"
        />
        <div
          class="css-j7qwjs"
        >
          <p
            class="chakra-text css-m26pul"
            max-width="180"
            word-wrap="break-word"
          >
            Don't Make Me Think
          </p>
          <div
            class="css-1709kwd"
          >
            <p
              class="chakra-text css-0"
            >
              Steve Krug
              ,
            </p>
            <p
              class="chakra-text css-0"
            >
              2000
            </p>
          </div>
          <p
            class="chakra-text css-0"
          >
            first
             Edition
          </p>
        </div>
      </div>
      <div
        class="css-109g2ia"
      >
        <p
          class="chakra-text css-blrpya"
        >
          4.5
          <span
            class="chakra-text css-1xibt3d"
          >
            /5
          </span>
        </p>
        <div
          class="css-16ol3ha"
        >
          <p
            class="chakra-text css-0"
          >
            SelfHelpBook
          </p>
          <p
            class="chakra-text css-0"
          >
            UX Design
          </p>
        </div>
      </div>
      <div
        class="css-1s14tyk"
      >
        <div
          class="css-1rb83ml"
        >
          <p
            class="chakra-text css-1619v3u"
          >
            In-Shelf
          </p>
        </div>
      </div>
      <div
        class="css-1s3ybnq"
      >
        <button
          aria-label="Heart icon"
          class="chakra-button css-1oer8ip"
          data-testid="handle-favorite"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="chakra-icon css-iz364c"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
            />
          </svg>
        </button>
        <button
          class="chakra-button css-4xx2wk"
          data-testid="redirect-preview"
          type="button"
        >
          Preview
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`My Book Shelf Favorites Search Should render correctly snapshot when user has no favorites 1`] = `
<div>
  <div
    class="css-6q2x0w"
  >
    <div
      class="css-1q7xe83"
    >
      <div
        class="css-dk70ac"
      >
        <img
          alt="Don't Make Me Think"
          data-nimg="1"
          decoding="async"
          fetchpriority="high"
          height="100"
          src="/_next/image?url=https%3A%2F%2Fi.ibb.co%2FQbQgVtG%2Fbook1.png&w=256&q=75"
          srcset="/_next/image?url=https%3A%2F%2Fi.ibb.co%2FQbQgVtG%2Fbook1.png&w=96&q=75 1x, /_next/image?url=https%3A%2F%2Fi.ibb.co%2FQbQgVtG%2Fbook1.png&w=256&q=75 2x"
          style="color: transparent;"
          width="75"
        />
        <div
          class="css-j7qwjs"
        >
          <p
            class="chakra-text css-m26pul"
            max-width="180"
            word-wrap="break-word"
          >
            Don't Make Me Think
          </p>
          <div
            class="css-1709kwd"
          >
            <p
              class="chakra-text css-0"
            >
              Steve Krug
              ,
            </p>
            <p
              class="chakra-text css-0"
            >
              2000
            </p>
          </div>
          <p
            class="chakra-text css-0"
          >
            first
             Edition
          </p>
        </div>
      </div>
      <div
        class="css-109g2ia"
      >
        <p
          class="chakra-text css-blrpya"
        >
          4.5
          <span
            class="chakra-text css-1xibt3d"
          >
            /5
          </span>
        </p>
        <div
          class="css-16ol3ha"
        >
          <p
            class="chakra-text css-0"
          >
            SelfHelpBook
          </p>
          <p
            class="chakra-text css-0"
          >
            UX Design
          </p>
        </div>
      </div>
      <div
        class="css-1s14tyk"
      >
        <div
          class="css-1rb83ml"
        >
          <p
            class="chakra-text css-1619v3u"
          >
            In-Shelf
          </p>
        </div>
      </div>
      <div
        class="css-1s3ybnq"
      >
        <button
          aria-label="Heart icon"
          class="chakra-button css-38t1hi"
          data-testid="handle-favorite"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="chakra-icon css-1k8v1e7"
            focusable="false"
            stroke-width="2"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
            />
          </svg>
        </button>
        <button
          class="chakra-button css-4xx2wk"
          data-testid="redirect-preview"
          type="button"
        >
          Preview
        </button>
      </div>
    </div>
  </div>
</div>
`;
